terraform/node_data_destroy.go

```diff
@@ -14,10 +15,26 @@ type NodeDestroyableDataResource struct {
 func (n *NodeDestroyableDataResource) EvalTree() EvalNode {
 	addr := n.ResourceInstanceAddr()
 
+	var providerSchema *ProviderSchema
+	// We don't need the provider, but we're calling EvalGetProvider to load the
+	// schema.
+	var provider providers.Interface
+
 	// Just destroy it.
 	var state *states.ResourceInstanceObject
-	return &EvalWriteState{
-		Addr:  addr.Resource,
-		State: &state, // state is nil here
+	return &EvalSequence{
+		Nodes: []EvalNode{
+			&EvalGetProvider{
+				Addr:   n.ResolvedProvider,
+				Output: &provider,
+				Schema: &providerSchema,
+			},
+			&EvalWriteState{
+				Addr:           addr.Resource,
+				State:          &state,
+				ProviderAddr:   n.ResolvedProvider,
```

Comment:
Since we've now repeated variants of this bug several times, maybe worth also having a check in `EvalWriteState.Eval` to explicitly check if `ProviderAddr` is set and return an error explicitly if not? That way this mistake will yield an error, rather than just quietly messing up the state.

Reply:
Good idea. I thought that would take some heavier refactoring, but apparently we have the provider in all other cases now.